File: frontend/src/utils/temporal.ts
/**
 * Temporal/bitemporal utility functions for handling valid_time and transaction_time ranges.
 *
 * @deprecated This module is deprecated. Use @/utils/localeDate for locale-aware date formatting.
 *
 * The backend now provides pre-formatted temporal data via computed fields:
 * - valid_time_formatted.lower_formatted
 * - transaction_time_formatted.lower_formatted
 *
 * For locale-aware formatting of backend dates, use:
 * - formatLocaleDate() - Format dates using browser locale
 * - formatLocaleDateTime() - Format datetimes using browser locale
 * - formatLocaleTemporalRange() - Format temporal ranges using browser locale
 *
 * This file is kept for backward compatibility during the migration period.
 */

/**
 * Type for temporal range values that can be either a string (PostgreSQL range)
 * or an object with start/lower properties (from some API responses).
 */
export type TemporalRange = string | { start?: string; lower?: string } | null | undefined;

/**
 * Date format style for locale-aware formatting.
 */
export type DateFormatStyle = "short" | "medium" | "long" | "full";

/**
 * Parse a temporal range value (string or object) and return the lower bound as a Date.
 *
 * @param range - Temporal range string (e.g., "[2026-01-15 10:00:00+00,)") or object with start/lower
 * @returns Date object representing the lower bound, or null if parsing fails
 *
 * @example
 * ```ts
 * parseTemporalRangeLowerBound("[2026-01-15 10:00:00+00,)") // => Date("2026-01-15T10:00:00+00:00")
 * parseTemporalRangeLowerBound({ start: "2026-01-15T10:00:00Z" }) // => Date("2026-01-15T10:00:00+00:00")
 * parseTemporalRangeLowerBound(null) // => null
 * ```
 */
export function parseTemporalRangeLowerBound(range: TemporalRange): Date | null {
  if (!range) return null;

  // Handle object format { start: string } or { lower: string }
  if (typeof range === "object") {
    const timestamp = range.start || range.lower;
    if (!timestamp) return null;
    const date = new Date(timestamp);
    return isNaN(date.getTime()) ? null : date;
  }

  // Handle string format (PostgreSQL range)
  return parseRangeLowerBound(range);
}

/**
 * Parse a PostgreSQL range string (e.g., "[2026-01-15 10:00:00+00,)")
 * and return the lower bound (start) timestamp as a Date.
 *
 * Range format: `[lower,upper)` where:
 * - `[` or `]` = inclusive bound
 * - `(` or `)` = exclusive bound
 * - Empty string for unbounded (e.g., ")" means infinity)
 *
 * @param rangeStr - PostgreSQL range string (e.g., "[2026-01-15 10:00:00+00,)")
 * @returns Date object representing the lower bound, or null if parsing fails
 *
 * @example
 * ```ts
 * parseRangeLowerBound("[2026-01-15 10:00:00+00,)") // => Date("2026-01-15T10:00:00+00:00")
 * parseRangeLowerBound(null) // => null
 * ```
 */
export function parseRangeLowerBound(rangeStr: string | null | undefined): Date | null {
  if (!rangeStr) return null;

  // PostgreSQL range format: "[lower,upper)" or "(lower,upper]"
  // Extract the timestamp between the opening bracket and the comma
  const commaIndex = rangeStr.indexOf(',');

  if (commaIndex === -1) return null;

  let timestamp = rangeStr.slice(1, commaIndex).trim();

  // Remove escaped quotes if present (JSON serialization of PostgreSQL ranges)
  timestamp = timestamp.replace(/\\"/g, '');

  // Check for infinity
  if (timestamp === '-infinity' || timestamp === 'infinity') {
    return null;
  }

  const date = new Date(timestamp);
  return isNaN(date.getTime()) ? null : date;
}

/**
 * Parse a PostgreSQL range string and return the upper bound (end) timestamp as a Date.
 *
 * @param rangeStr - PostgreSQL range string (e.g., "[2026-01-15 10:00:00,)")
 * @returns Date object representing the upper bound, or null if unbounded or parsing fails
 *
 * @example
 * ```ts
 * parseRangeUpperBound("[2026-01-15 10:00:00,2026-02-15 10:00:00+00)") // => Date("2026-02-15T10:00:00+00:00")
 * parseRangeUpperBound("[2026-01-15 10:00:00,)") // => null (unbounded)
 * ```
 */
export function parseRangeUpperBound(rangeStr: string | null | undefined): Date | null {
  if (!rangeStr) return null;

  // PostgreSQL range format: "[lower,upper)" or "(lower,upper]"
  // Extract the timestamp between the comma and the closing bracket
  const commaIndex = rangeStr.indexOf(',');

  if (commaIndex === -1) return null;

  let timestamp = rangeStr.slice(commaIndex + 1, -1).trim();

  // Remove escaped quotes if present (JSON serialization of PostgreSQL ranges)
  timestamp = timestamp.replace(/\\"/g, '');

  // Check for empty (unbounded) or infinity
  if (!timestamp || timestamp === '-infinity' || timestamp === 'infinity') {
    return null; // Unbounded upper bound
  }

  const date = new Date(timestamp);
  return isNaN(date.getTime()) ? null : date;
}

/**
 * Check if a range string has an unbounded upper bound (i.e., is currently valid).
 *
 * @param rangeStr - PostgreSQL range string (e.g., "[2026-01-15 10:00:00,)")
 * @returns true if the upper bound is unbounded (currently valid), false otherwise
 *
 * @example
 * ```ts
 * isRangeUnbounded("[2026-01-15 10:00:00,)") // => true
 * isRangeUnbounded("[2026-01-15 10:00:00,2026-02-15 10:00:00)") // => false
 * ```
 */
export function isRangeUnbounded(rangeStr: string | null | undefined): boolean {
  if (!rangeStr) return false;
  const commaIndex = rangeStr.indexOf(',');

  if (commaIndex === -1) return false;

  let upperBound = rangeStr.slice(commaIndex + 1, -1).trim();

  // Remove escaped quotes if present (JSON serialization of PostgreSQL ranges)
  upperBound = upperBound.replace(/\\"/g, '');

  // Empty or infinity means unbounded
  return !upperBound || upperBound === '-infinity' || upperBound === 'infinity';
}

/**
 * Format a temporal range for display as a locale-aware date string.
 * Handles both valid_time and transaction_time ranges in string or object format.
 * Uses the browser's locale settings for automatic formatting.
 *
 * @param range - Temporal range string or object
 * @param style - Date format style (default: "medium")
 * @returns Formatted date string using browser locale, or "Unknown" if parsing fails
 *
 * @example
 * ```ts
 * formatTemporalDate("[2026-01-15 10:00:00+00,)") // => "Jan 15, 2026" (en-US locale)
 * formatTemporalDate("[2026-01-15 10:00:00+00,)", "short") // => "1/15/26" (en-US locale)
 * formatTemporalDate("[2026-01-15 10:00:00+00,)", "long") // => "January 15, 2026" (en-US locale)
 * formatTemporalDate(null) // => "Unknown"
 * ```
 */
export function formatTemporalDate(
  range: TemporalRange,
  style: DateFormatStyle = "medium"
): string {
  const date = parseTemporalRangeLowerBound(range);
  if (!date) return "Unknown";

  // Use locale-aware formatting with the browser's default locale
  return date.toLocaleDateString(undefined, { dateStyle: style });
}

/**
 * Format a temporal range for display as a locale-aware date string with time.
 * Includes both date and time portions.
 *
 * @param range - Temporal range string or object
 * @param style - Date format style (default: "medium")
 * @returns Formatted datetime string using browser locale, or "Unknown" if parsing fails
 *
 * @example
 * ```ts
 * formatTemporalDateTime("[2026-01-15 10:30:00+00,)") // => "Jan 15, 2026, 10:30 AM" (en-US locale)
 * formatTemporalDateTime("[2026-01-15 10:30:00+00,)", "short") // => "1/15/26, 10:30 AM" (en-US locale)
 * formatTemporalDateTime(null) // => "Unknown"
 * ```
 */
export function formatTemporalDateTime(
  range: TemporalRange,
  style: DateFormatStyle = "medium"
): string {
  const date = parseTemporalRangeLowerBound(range);
  if (!date) return "Unknown";

  // Use locale-aware formatting with date and time
  return date.toLocaleString(undefined, {
    dateStyle: style,
    timeStyle: "short"
  });
}

/**
 * Format a PostgreSQL range string for display as a locale-aware date label.
 * Handles both valid_time and transaction_time ranges.
 *
 * @param rangeStr - PostgreSQL range string (e.g., "[2026-01-15 10:00:00+00,)")
 * @param style - Date format style (default: "medium")
 * @returns Formatted date string using browser locale, or "Unknown" if parsing fails
 *
 * @example
 * ```ts
 * formatRangeDate("[2026-01-15 10:00:00+00,)") // => "Jan 15, 2026" (en-US locale)
 * formatRangeDate("[2026-01-15 10:00:00+00,)", "short") // => "1/15/26" (en-US locale)
 * formatRangeDate(null) // => "Unknown"
 * ```
 */
export function formatRangeDate(
  rangeStr: string | null | undefined,
  style: DateFormatStyle = "medium"
): string {
  return formatTemporalDate(rangeStr, style);
}

File: backend/app/core/temporal.py
"""Temporal data formatting utilities for PostgreSQL TSTZRANGE fields.

This module provides utilities to convert PostgreSQL TSTZRANGE database
values into display-ready formats for API responses. This keeps the
frontend decoupled from database-specific serialization details.
"""

from datetime import datetime
from typing import Any


def format_temporal_range_for_api(
    range_value: Any,
) -> dict[str, str | None]:
    """Convert PostgreSQL TSTZRANGE to display-ready format.

    Extracts temporal range information and returns both raw ISO timestamps
    and formatted display strings. This allows the frontend to simply
    display pre-formatted dates without parsing PostgreSQL range syntax.

    Args:
        range_value: TSTZRANGE value from database (can be range object,
                     pre-formatted string, or None)

    Returns:
        Dictionary with:
            - lower: ISO 8601 timestamp of lower bound or None
            - upper: ISO 8601 timestamp of upper bound or None (None if unbounded)
            - lower_formatted: Human-readable lower bound date (e.g., "Jan 15, 2026")
            - upper_formatted: "Present" if unbounded, or formatted upper date
            - is_currently_valid: True if the range is currently valid (unbounded upper)

    Examples:
        >>> format_temporal_range_for_api("[\\"2026-01-15 10:00:00+00\\",)")
        {
            "lower": "2026-01-15T10:00:00+00:00",
            "upper": None,
            "lower_formatted": "January 15, 2026",
            "upper_formatted": "Present",
            "is_currently_valid": True
        }

        >>> format_temporal_range_for_api("[\\"2026-01-15 10:00:00+00\\",\\"2026-02-15 10:00:00+00\\")]")
        {
            "lower": "2026-01-15T10:00:00+00:00",
            "upper": "2026-02-15T10:00:00+00:00",
            "lower_formatted": "January 15, 2026",
            "upper_formatted": "February 15, 2026",
            "is_currently_valid": False
        }
    """
    if not range_value:
        return _get_empty_range_dict()

    # Handle pre-formatted ISO string (from ProjectRead.convert_range_to_iso)
    if isinstance(range_value, str):
        # Check if it's already an ISO timestamp (not a range string)
        if range_value.startswith("20") and not range_value.startswith("["):
            try:
                dt = datetime.fromisoformat(range_value.replace("Z", "+00:00"))
                return {
                    "lower": range_value,
                    "upper": None,
                    "lower_formatted": _format_datetime(dt),
                    "upper_formatted": "Present",
                    "is_currently_valid": True,
                }
            except ValueError:
                return _get_empty_range_dict()

        # Handle PostgreSQL range string format: ["2026-01-15 10:00:00+00",)
        return _parse_postgresql_range_string(range_value)

    # Handle PostgreSQL range object (with .lower and .upper attributes)
    if hasattr(range_value, "lower") and range_value.lower:
        lower_dt = datetime.fromisoformat(
            range_value.lower.isoformat()
        ) if isinstance(range_value.lower, datetime) else range_value.lower

        upper_dt = None
        if hasattr(range_value, "upper") and range_value.upper:
            upper_dt = datetime.fromisoformat(
                range_value.upper.isoformat()
            ) if isinstance(range_value.upper, datetime) else range_value.upper

        return {
            "lower": lower_dt.isoformat(),
            "upper": upper_dt.isoformat() if upper_dt else None,
            "lower_formatted": _format_datetime(lower_dt),
            "upper_formatted": _format_datetime(upper_dt) if upper_dt else "Present",
            "is_currently_valid": upper_dt is None,
        }

    return _get_empty_range_dict()


def _parse_postgresql_range_string(range_str: str) -> dict[str, str | None]:
    """Parse PostgreSQL TSTZRANGE string format.

    Expected format: ["lower",upper") or ["lower",upper] etc.
    JSON serialization adds escaped quotes: [\\"lower\\",upper)

    Args:
        range_str: PostgreSQL range string (e.g., '[\\"2026-01-15 10:00:00+00\\",)')

    Returns:
        Dictionary with temporal range information
    """
    if not range_str or "," not in range_str:
        return _get_empty_range_dict()

    try:
        # Find comma and extract bounds
        comma_index = range_str.index(",")

        # Extract lower bound (after opening bracket, before comma)
        lower_str = range_str[1:comma_index].strip()

        # Remove escaped quotes if present (JSON serialization artifact)
        lower_str = lower_str.replace('\\"', "")

        # Extract upper bound (after comma, before closing bracket)
        # PostgreSQL ranges can end with ) for exclusive or ] for inclusive
        # Also handle case where range might be followed by ] (JSON array)
        upper_str = range_str[comma_index + 1:].strip()

        # Find the actual closing bracket (either ) or ])
        if upper_str.endswith(")]"):
            upper_str = upper_str[:-2]  # Remove )]
        elif upper_str.endswith(")") or upper_str.endswith("]"):
            upper_str = upper_str[:-1]  # Remove ) or ]

        upper_str = upper_str.strip()
        upper_str = upper_str.replace('\\"', "")

        # Check for unbounded (empty string or infinity)
        is_unbounded = not upper_str or upper_str in ("-infinity", "infinity")

        # Parse lower bound
        lower_dt = datetime.fromisoformat(lower_str) if lower_str else None

        # Parse upper bound if bounded
        upper_dt = None
        if not is_unbounded and upper_str:
            upper_dt = datetime.fromisoformat(upper_str)

        return {
            "lower": lower_dt.isoformat() if lower_dt else None,
            "upper": upper_dt.isoformat() if upper_dt else None,
            "lower_formatted": _format_datetime(lower_dt) if lower_dt else "Unknown",
            "upper_formatted": "Present" if is_unbounded else (
                _format_datetime(upper_dt) if upper_dt else "Unknown"
            ),
            "is_currently_valid": is_unbounded,
        }
    except (ValueError, IndexError):
        return _get_empty_range_dict()


def _format_datetime(dt: datetime | None) -> str:
    """Format datetime for display.

    Uses a consistent format across all API responses.
    Frontend can display this directly without further processing.

    Args:
        dt: datetime object or None

    Returns:
        Formatted date string (e.g., "January 15, 2026") or "Unknown"
    """
    if not dt:
        return "Unknown"

    # Format: "January 15, 2026"
    return dt.strftime("%B %d, %Y")


def _get_empty_range_dict() -> dict[str, str | None]:
    """Return dictionary for empty/invalid temporal ranges."""
    return {
        "lower": None,
        "upper": None,
        "lower_formatted": "Unknown",
        "upper_formatted": "Unknown",
        "is_currently_valid": False,
    }
